src/s3path.jl

```diff
@@ -1,14 +1,31 @@
-struct S3Path <: AbstractPath
+struct S3Path{A<:AbstractAWSConfig} <: AbstractPath
     segments::Tuple{Vararg{String}}
     root::String
     drive::String
     isdirectory::Bool
-    config::AWSConfig
+    config::A
+
+    # this method ensures that there is proper conversion to Tuple{Vararg{String}} for first arg,
+    # which can be a little tempremental if you don't explicitly handle it
+    function S3Path{A}(
+        segments,
+        root::AbstractString,
+        drive::AbstractString,
+        isdirectory::Bool,
+        config::A
+    ) where {A<:AbstractAWSConfig}
+        new{typeof(config)}(tuple((String(s) for s ∈ segments)...), root, drive, isdirectory, config)
```

Comment:
Just want to cc @rofinn here since he is the mastermind behind this pathing.
Not sure if opinions are had here or not.

Reply:
The issue is that `Tuple{Vararg{String}}` is much stricter than other types (e.g. `Vector`) with conversions, so one has to be careful with e.g. `Tuple{Vararg{SubString}}`.  I considered changing the type of the field, since it's not even a concrete type, but I noticed that this is the same field type as those in `FilePathsBase` so I figured it was best to leave it alone and just make sure it wasn't going to crash.